test/units/testsuite-46.sh

```diff
@@ -170,6 +168,119 @@ if ! systemd-detect-virt -cq ; then
     homectl remove test-user2
 fi
 
+# blob directory tests
+# See docs/USER_RECORD_BLOB_DIRS.md
+checkblob() {
+        test -f "/var/cache/systemd/home/blob-user/$1"
+        stat -c "%u %#a" "/var/cache/systemd/home/blob-user/$1" | grep "^0 0644"
+        test -f "/home/blob-user/.identity-blob/$1"
+        stat -c "%u %#a" "/home/blob-user/.identity-blob/$1" | grep "12345 0644"
+
+        diff "/var/cache/systemd/home/blob-user/$1" "$2"
+        diff "/var/cache/systemd/home/blob-user/$1" "/home/blob-user/.identity-blob/$1"
+}
+
+mkdir /tmp/blob1 /tmp/blob2
+echo data1 blob1 > /tmp/blob1/test1
+echo data1 blob2 > /tmp/blob2/test1
+echo data2 blob1 > /tmp/blob1/test2
+echo data2 blob2 > /tmp/blob2/test2
+echo invalid filename > /tmp/blob1/файл
+ln -s /tmp/blob1/test1 /tmp/blob1/symlink
+echo data3 > /tmp/external-test3
+echo avatardata > /tmp/external-avatar
+ln -s /tmp/external-avatar /tmp/external-avatar-lnk
+
+# create w/ prepopulated blob dir
+NEWPASSWORD=EMJuc3zQaMibJo homectl create blob-user --uid=12345 \
```

Comment:
CentOS CI Arch linux OOM-kills homework here...

So I either have a memory leak somewhere subtle, or the test needs to have the amount of memory allocated to the VM bumped

Is there some way to run the tests w/ valgrind? Am I going to have to do it locally (I'm not really set up for it...)? Spot any memory leaks I missed?

Reply:
there's a CI test with address and memory sanitizer. It's generally pretty much as good as valgrind, and runs our whole testsuite like that. (`CentOS CI (Arch Linux + sanitizers)`)

Reply:
Well that's interesting. One of the files we try to copy into the blob directory is bigger than the 64M limit now enforced by `sha256_fd`. Now of course that makes no sense (each file in there should only be a couple of bytes big!) so I'll need to investigate further.

But it explains the OOM: I didn't previously pass a limit into `copy_bytes`, so I guess that `copy_bytes` would allocate enough in response to this apparently huge file that we get OOM killed. I now pass a maximum size into `copy_bytes`, and we no longer get OOM killed.

Reply:
Another part of the mystery of the OOM: Turns out we were getting OOM-killed during resize. I neglected to set disk size limits on `homectl create`, so when homed tried to grow the home dir it got OOM killed

Reply:
```
Blob test1 size is 12B, making the new total dir size 86.0T
```

:scream:

Reply:
Tests are passing now